test/common/buffer/buffer_speed_test.cc

```diff
@@ -0,0 +1,221 @@
+  for (auto _ : state) {
+    constexpr uint64_t NumSlices = 2;
+    Buffer::RawSlice slices[NumSlices];
+    uint64_t slices_used = buffer.reserve(state.range(0), slices, NumSlices);
+    uint64_t bytes_to_commit = 0;
+    for (uint64_t i = 0; i < slices_used; i++) {
+      bytes_to_commit += static_cast<uint64_t>(slices[i].len_);
+    }
+    buffer.commit(slices, slices_used);
+    buffer.drain(bytes_to_commit);
+  }
+  benchmark::DoNotOptimize(buffer.length());
+}
+BENCHMARK(BufferReserveCommit)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+// Test the reserve+commit cycle, for the common case where the reserved space is
+// only partially used (and therefore the commit size is smaller than the reservation size).
+static void BufferReserveCommitPartial(benchmark::State& state) {
+  Buffer::OwnedImpl buffer;
+  for (auto _ : state) {
+    constexpr uint64_t NumSlices = 2;
+    Buffer::RawSlice slices[NumSlices];
+    uint64_t slices_used = buffer.reserve(state.range(0), slices, NumSlices);
+    ASSERT(slices_used > 0);
+    // Commit one byte from the first slice and nothing from any subsequent slice.
+    uint64_t bytes_to_commit = 1;
+    slices[0].len_ = bytes_to_commit;
+    buffer.commit(slices, 1);
+    buffer.drain(bytes_to_commit);
+  }
+  benchmark::DoNotOptimize(buffer.length());
+}
+BENCHMARK(BufferReserveCommitPartial)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+// Test buffer search, for the simple case where there are no partial matches for
+// the pattern in the buffer.
+static void BufferSearch(benchmark::State& state) {
+  const std::string Pattern(16, 'b');
+  std::string data;
+  data.reserve(state.range(0) + Pattern.length());
+  data += std::string(state.range(0), 'a');
+  data += Pattern;
+
+  const absl::string_view input(data);
+  Buffer::OwnedImpl buffer(input);
+  ssize_t result = 0;
+  for (auto _ : state) {
+    result += buffer.search(Pattern.c_str(), Pattern.length(), 0);
+  }
+  benchmark::DoNotOptimize(result);
+}
+BENCHMARK(BufferSearch)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+// Test buffer search, for the more challenging case where there are many partial matches
+// for the pattern in the buffer.
+static void BufferSearchPartialMatch(benchmark::State& state) {
+  const std::string Pattern(16, 'b');
+  const std::string PartialMatch("babbabbbabbbbabbbbbabbbbbbabbbbbbbabbbbbbbba");
+  std::string data;
+  size_t num_partial_matches = 1 + state.range(0) / PartialMatch.length();
+  data.reserve(state.range(0) * num_partial_matches + Pattern.length());
+  for (size_t i = 0; i < num_partial_matches; i++) {
+    data += PartialMatch;
+  }
+  data += Pattern;
+
+  const absl::string_view input(data);
+  Buffer::OwnedImpl buffer(input);
+  ssize_t result = 0;
+  for (auto _ : state) {
+    result += buffer.search(Pattern.c_str(), Pattern.length(), 0);
+  }
+  benchmark::DoNotOptimize(result);
+}
+BENCHMARK(BufferSearchPartialMatch)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+} // namespace Envoy
+
```

Comment:
looks like there are some functions we're not testing which are in upstream envoy.  linearize, prepend etc.   They're not in the common path but it'd be good to still cover them.

Reply:
i'm having a little trouble understanding github's comment history, but can you update the PR description to include stats from your new tests, you haven't already done that?

Reply:
Ok, I added the latest benchmark output to the PR description.